readme.md
Shattered Lands
===============

The planet you landed on has deep chasms and is full of rocks and cliffs. Traveling and moving resources will be a challenge. Good thing that you have some technologies to help you.

Gameplay
---------

This mod changes map generation to make the map more difficult to navigate. It generates abyss tiles and adds more environmental hazards. It also adds some modes of transportation, namely a grappling hook and spidertron logistics.

When starting a new game, select the Shattered Lands preset!

The map generation creates deep chasms with narrow bridges in between them. Some bridges are crumbled, but can still be crossed with a spidertron. The abyss tiles can not be filled with landfill. The map also contains more cliffs and rocks. The map is labyrinthine, has narrow choke points and space can be scarce at times. Transporting resources is a big challenge here, rail networks are basically impossible. You'll have to think about where you build which production chain, in order to use space optimally and keep transportation costs low. On the other hand, defending against biters is easier, you usually only have to defend narrow choke points instead of building long walls.

Some movement options are provided via other mods which are installed automatically by this mod:

* __Jump and Swing__ - My reworked grappling hook and jump mod. This is the main way of moving the character around the map.
* __Spidertron Patrols__ - Allows automating resource movement via spidertrons. Main method of getting resources through difficult terrain, but only for short segments or low throughput. Makes a small spidertron variant available at blue science.
* __Lex's Aircraft__ - Aircrafts that are controlled like spidertrons and hence can be automated. Useful for transporting resources over ravines, but need to be fueled via logistic bots.
* __Bulk Teleportation__ - Allows teleportation of resources at a steep electricity cost. Strongly nerfed compared to the original version and only available very late. It costs around 10MW to connect a yellow belt. Basically the replacement for trains in the late game.

![Short Demo](https://raw.githubusercontent.com/BlueOne/factorio-shattered-lands/main/.web/shattered-lands-transportation.mp4)

Hand-Picked Map Seeds
-----------------------

Here are some hand-picked seeds: 2965077726, 3347003385, 273310745, 2745948135, 2959005515, 699788883, 2030477123. The mod doesn't make sure the game is beatable: in very rare cases, oil might not be reachable or other resources might only be available at a large distance outside the starting area. On the other hand, one can sometimes roll a map with very good terrain near the starting area, so the game wouldn't be too different from a normal playthrough. The hand-picked seeds avoid both these issues.

Compatibility / Optional Mods
-----------------------------

__Airplane Mods:__ The mod comes with Lex's airships mod for logistics, for personal transportation you are intended to use the grappling hook or the airplanes. Aircraft Realism is not suggested; you get stuck when you land on the abyss.

__Map Generation Mods:__ Compatible with alien biomes (but you don't really need it, the terrain already looks different from the base game). Possibly not compatible with other mods that change map generation.

__Character mods:__ Since this mod relies on my grappling hook mod, it is incompatible with number of mods that influence the character, in particular the jetpack mod is incompatible at the moment.

__Recipe Mods / Overhaul Mods:__ In theory compatible with many overhauls, but I recommend a short game length. Recommendations

* Production Scrap 2 (install Production Scrap 2 Icon Fix by nIronwolf).
* K2
* AAI industry

__Mining Mods:__ When a patch is directly next to abyss, you can sometimes not mine the entire patch with normal mining drills. With Pepperming Mining you can make bots do the job. Of course you can also use a large-area drill mod.

Acknowledgments
------------

Thanks to:

* Sparkletron, whose terraforming mod is the basis for this mod.
* The authors of the transportation mods this one depends on: Xorimuth - Spidertron Patrols, Lex - Lex's Aircraft, dorfl - Bulk Teleporters

For Mod Devs
------------

Instead of using the standard out-of-map tile, this mod uses a copied tile called abyss, this is to make it clear that it can be crossed via movement mods such as the jump, grappling hook or jetpack.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2023.09.30
  Info:
    - Modify menu simulations from base game.
    - Replace aircrafts mod with Lex's airships mod.
    - Fix bug where the inventory of item teleporters wasn't adjusted properly.
    - Optionally support Production Scrap 2

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2023.09.30
  Info:
    - Initial Release


scripts/util.lua

local Util = {}
local util = require("__core__/lualib/util.lua")


-- Ruins spawning etc.
-----------------------------------------------------------


function Util.custom_insert_safe (entity, item_dict)
  if not (entity and entity.valid and item_dict) then return end

  local insert = entity.insert
  for name, count in pairs (item_dict) do
    -- check if the item exists
    local item_type = Util.cached_item_type(name)
    if item_type then
      insert{name = name, count = count}
    end
  end
end 

function Util.insert_fluid_safe(entity, fluid_dict)
  if not (entity and entity.valid and fluid_dict) then return end
  local fluids = game.fluid_prototypes
  local insert = entity.insert_fluid
  for name, amount in pairs (fluid_dict) do
    if fluids[name] then
      insert{name = name, amount = amount}
    else
      log("Fluid to insert not valid: " .. name)
    end
  end
end


function Util.insert_loot(containers, loot_table)
  -- check if all containers are valid
  local valid_containers = {}
  for _, container in pairs(containers) do
    if container.valid then table.insert(valid_containers, container) end
  end

  for _, item_stack in pairs(loot_table) do
    local container = Util.random_choice_array(containers)
    container.insert(item_stack)
  end
end

-- Cached API calls
-----------------------------------------------------------

function Util.entity_type(name)
  if game.entity_prototypes[name] then return game.entity_prototypes[name].type end
  return false
end

function Util.item_type(name)
  return game.item_prototypes[name] and game.item_prototypes[name].type or false
end

function Util.remnant_names(name)
  local prototype = game.entity_prototypes[name]
  if not prototype then
    return false
  else
    return Util.keys(prototype.corpses or {})
  end
end

function Util.entity_prototype_selection_box(name)
  local prototype = game.entity_prototypes[name]
  if not prototype then return false end
  return prototype.selection_box
end

function Util.cached(fn, cache)
  local f = function(input)
    local cached_result = cache[input]
    if cached_result ~= nil then return cached_result end
    local result = fn(input)
    cache[input] = result
    return result
  end
  return f
end

local entity_selection_box_cache = {}
Util.cached_entity_selection_box = Util.cached(Util.entity_prototype_selection_box, entity_selection_box_cache)

local entity_type_cache = {}
Util.cached_entity_type = Util.cached(Util.entity_type, entity_type_cache)

local item_type_cache = {}
Util.cached_item_type = Util.cached(Util.item_type, item_type_cache)

local remnant_cache = {}
Util.cached_remnant_names = Util.cached(Util.remnant_names, remnant_cache)



-- Geometry
-----------------------------------------------------------

function Util.random_point_in_rect(rect, generator)
  generator = generator or math.random
  local x
  local y
  x = generator(rect.left_top.x, rect.right_bottom.x)
  y = generator(rect.left_top.y, rect.right_bottom.y)
  return { x = x, y = y }
end


function Util.is_proper_rect(rect)
  return rect.left_top.x < rect.right_bottom.x and rect.left_top.y < rect.right_bottom.y
end


function Util.intersect_aabb(rect1, rect2)
  for _, k in pairs({"x", "y"}) do
    if rect1.right_bottom[k] < rect2.left_top[k] or rect2.right_bottom[k] < rect1.left_top[k] then return false end
  end
  return true
end

function Util.direction_to_orientation(direction)
  if direction == defines.direction.north then
      return 0
  elseif direction == defines.direction.northeast then
      return 0.125
  elseif direction == defines.direction.east then
      return 0.25
  elseif direction == defines.direction.southeast then
      return 0.375
  elseif direction == defines.direction.south then
      return 0.5
  elseif direction == defines.direction.southwest then
      return 0.625
  elseif direction == defines.direction.west then
      return 0.75
  elseif direction == defines.direction.northwest then
      return 0.875
  end
  return 0
end

function Util.orientation_to_direction(orientation)
  orientation = (orientation + 0.0625) % 1
  if orientation <= 0.125 then
    return defines.direction.north
  elseif orientation <= 0.25 then
    return defines.direction.northeast
  elseif orientation <= 0.375 then
    return defines.direction.east
  elseif orientation <= 0.5 then
    return defines.direction.southeast
  elseif orientation <= 0.625 then
    return defines.direction.south
  elseif orientation <= 0.75 then
    return defines.direction.southwest
  elseif orientation <= 0.875 then
    return defines.direction.west
  else
    return defines.direction.northwest
  end
end



-- Remote interfaces
-----------------------------------------------------------

function Util.expose_remote_interface(module, name, function_names)
  local functions = {}
  for _, k in pairs(function_names) do
    local v = module[k]
    if type(v) == "function" then
      functions[k] = v
    end
  end
  remote.add_interface(name, functions)
end

function Util.expose_remote_interface_all(module, name)
  local functions = {}
  for k, v in pairs(module) do
    if type(v) == "function" then
      functions[k] = v
    end
  end
  Util.remote_add_interface(name, functions)
end


-- Stochastics
-----------------------------------------------------------

function Util.sample_poisson(lambda, generator)
  local r = generator()
  local el = math.exp(-lambda)
  local i_factorial = 1
  local lambda_pw = 1
  for i = 0, 4 do
      if i > 0 then i_factorial = i_factorial * i end
      local p = el * lambda_pw / i_factorial
      lambda_pw = lambda_pw * lambda
      if r >= 1 - p then return i end
      r = r + p
  end
  return 5
end



-- Tables
-----------------------------------------------------------

Util.compare = util.table.compare
Util.deepcopy = util.table.deepcopy
Util.copy = util.table.deepcopy
Util.merge = util.merge

function Util.table_size(t)
  local size = 0
  for _, _ in pairs(t) do
    size = size + 1
  end
  return size
end

function Util.keys(t)
  local keys = {}
  for k, _ in pairs(t) do
    table.insert(keys, k)
  end
  return keys
end

function Util.values(t)
  local values = {}
  for _, v in pairs(t) do
    table.insert(values, v)
  end
  return values
end

function Util.table_contains(t, v)
  for _, v1 in pairs(t) do
    if v1 == v then return true end
  end
  return false
end

-- Example calls:
-- find(elem, list)
-- find(elem, list, equal_func)
-- find(list, identifier_func)
-- Returns key, value or false.

function Util.find(arg1, arg2, arg3)
  -- find(list, identifier_func)
    if type(arg1) == "table" and type(arg2) == "function" then
      for k, v in pairs(arg1) do
        if arg2(v) then
          return k, v
        end
      end
      return false
    end
  
    -- find(elem, list)
    if not arg3 then
      if type(arg1) == "table" then
        arg3 = Util.compare
      else
        for k, other in pairs(arg2) do
          if arg1 == other then
            return k, other
          end
        end
        return false
      end
    end
  
    -- find(elem, list, equal_func)
    for k, other in pairs(arg2) do
      if arg3(arg1, other) then
        return k, other
      end
    end
    return false
  end
  

function Util.find_minimum(list, lessthan_func)
	if #list == 0 then return 0 end
	local index = 1
	local min_value = list[index]
	for i, value in ipairs(list) do
		if lessthan_func then
			if lessthan_func(value, min_value) then
				index = i
				min_value = value
			end
		else
			if value < min_value then
				index = i
				min_value = value
			end
		end
	end
	return index, min_value
end


function Util.increment(t, k, v)
	if not t[k] then
		t[k] = v or 1
	else
		t[k] = t[k] + (v or 1)
	end
end

function Util.decrement(t, k, v)
	if not t[k] then
		t[k] = -(v or 1)
	else
		t[k] = t[k] - (v or 1)
	end
end


function Util.all_wrong(t)
  for _, v in pairs(t) do
    if v then return false end
  end
  return true
end

function Util.remove_from_table(list, item)
  local index = 0
  for _,_item in ipairs(list) do
      if item == _item then
          index = _
          break
      end
  end
  if index > 0 then
      table.remove(list, index)
  end
end



-- Randomness
function Util.random_choice_array(t, generator)
  generator = generator or math.random
  local count = #t
  local num = generator(count)
  return t[num]
end

function Util.random_choice_table(t, generator)
  generator = generator or math.random
  local count = Util.table_size(t)
  local num = generator(count)
  local key = next(t)
  while num > 1 do
    key = next(t, key)
    num = num - 1
  end

  return t[key], key
end

return util.merge{util, Util, util.table}


data-final-fixes.lua

-- Fix transitions with other tiles
for _, tile in pairs(data.raw.tile) do
    if tile.transitions then
        for _, transition in pairs(tile.transitions) do
            local out_of_map_found = false
            local abyss_found = false
            if transition.to_tiles then
                for _, tile_name in pairs(transition.to_tiles) do
                    if tile_name == "out-of-map" then
                        out_of_map_found = true
                    end
                    if tile_name == "abyss" then
                        abyss_found = true
                    end
                    if out_of_map_found and not abyss_found then
                        table.insert(transition.to_tiles, "abyss")
                        transition.transition_group = 2
                    end
                end
            end
        end
    end
end

data.raw.tile.abyss.layer = 0

if data.raw.car["hcraft-entity"] then
    local _, hovercraft_layer = next(data.raw.car["hcraft-entity"].collision_mask)
    table.insert(data.raw.tile.abyss.collision_mask, hovercraft_layer)
end


data.lua
local Util = require("scripts.util")

local noise = require("noise")
local terraformer = require("scripts.mapGenNoise")
local tne = noise.to_noise_expression

-- Rocks
------------------------------------------------------------------------

local function rock_peak_old(rectangle, influence)
    local peak =
    {
        noise_layer = "rocks",
        noise_octaves_difference = -2,
        noise_persistence = 0.75,
        influence = influence or 1
    }

    if rectangle ~= nil then
        local aux_center = (rectangle[2][1] + rectangle[1][1]) / 2
        local aux_range = math.abs(rectangle[2][1] - rectangle[1][1]) / 2
        local water_center = (rectangle[2][2] + rectangle[1][2]) / 2
        local water_range = math.abs(rectangle[2][2] - rectangle[1][2]) / 2

        peak["aux_optimal"] = aux_center
        peak["aux_range"] = aux_range
        peak["aux_max_range"] = water_range + 0.05

        peak["water_optimal"] = water_center
        peak["water_range"] = water_range
        peak["water_max_range"] = water_range + 0.05
    end
    return peak
end

local rock_peak = {
    noise_layer = "rocks",
    noise_octaves_difference = -2,
    noise_persistence = 0.75,
}


local function rock_autoplace_settings(coverage, max_probability, order_suffix, tile_restriction)
    return
    {
        order = "a[doodad]-a[rock]-" .. order_suffix,
        coverage = coverage,
        sharpness = 0.9,
        max_probability = max_probability,
        peaks = { rock_peak },
        tile_restriction = tile_restriction
    }
end


-- local normal_rock_peaks = {rock_peak{{0, 0.3}, {1, 1}}, --[[rock_peak({{0.5, 0.}, {1, 0.2}}, 1)]]}
-- local sand_rock_peaks = {rock_peak{{0, 0}, {0.4, 0.2}}}

local coverage = { ["rock-big"] = 0.07, ["rock-huge"] = 0.04, ["sand-rock-big"] = 0.1 }
local max_probability = { ["rock-big"] = 0.2, ["rock-huge"] = 0.2, ["sand-rock-big"] = 0.1 }

alien_biomes = alien_biomes or nil
if not alien_biomes then
    local sand_tiles = { "sand-1", "sand-2", "sand-3", "red-desert-0", "red-desert-1", "red-desert-2", "red-desert-3" }
    local not_sand_tiles = {}
    for _, tile in pairs(data.raw.tile) do
        if not Util.find(tile.name, sand_tiles) then
            table.insert(not_sand_tiles, tile.name)
        end
    end

    -- for _, prototype in pairs(data.raw["simple-entity"]) do
    --     if string.find(prototype.name, "rock") then
    --         prototype.autoplace.peaks = Util.copy(rock_peak)
    --         prototype.
    --     end
    -- end
    for _, rock_name in pairs({ "rock-big", "rock-huge", "sand-rock-big" }) do
        local prototype = data.raw["simple-entity"][rock_name]
        local cvg = coverage[rock_name]
        local max_pr = max_probability[rock_name]
        prototype.autoplace = rock_autoplace_settings(cvg, max_pr, "b[" .. rock_name .. "]", not_sand_tiles)
    end
else
    for _, rock_name in pairs({ "rock-big", "rock-huge", "sand-rock-big" }) do
        local cvg = coverage[rock_name]
        local max_pr = max_probability[rock_name]
        for name, _ in pairs({
            tan = { 193, 162, 127 },
            white = { 255, 255, 255 },
            grey = { 177, 183, 187 },
            black = { 135, 135, 135 },
            purple = { 169, 177, 239 },
            red = { 185, 107, 105 },
            violet = { 165, 107, 161 },
            dustyrose = { 180, 148, 137 },
            cream = { 234, 216, 179 },
            brown = { 162, 117, 88 },
            beige = { 178, 164, 138 },
            aubergine = { 126, 115, 156 }
        }) do
            local prototype_name = rock_name .. '-' .. name
            local prototype = data.raw["simple-entity"][prototype_name]
            if prototype then
                prototype.autoplace.coverage = cvg
                prototype.autoplace.sharpness = 0.9
                prototype.autoplace.max_probability = max_pr
                prototype.autoplace.peaks = { rock_peak }
            end
        end
    end
end


-- Abyss
------------------------------------------------------------------------

data:extend({
    {
        type = "autoplace-control",
        name = "abyss",
        order = "z-a",
        category = "terrain",
        richness = true,
    },
    {
        type = "noise-expression",
        name = "control-setting:abyss:frequency:multiplier",
        expression = tne(1)
    },
    {
        type = "noise-expression",
        name = "control-setting:abyss:bias",
        expression = tne(0)
    },

})

-- Noise Expression
data:extend
{
    {
        type = "noise-expression",
        name = "abyss-elevation",
        expression = noise.define_noise_function(
            function(x, y, tile, _map)
                -- Hot gargage hack to make this independent of water properties
                -- TODO: Link to settings sliders
                local map = Util.copy(_map)
                map.segmentation_multiplier = noise.get_control_setting("abyss").frequency_multiplier
                map.wlc_elevation_offset = noise.get_control_setting("abyss").size_multiplier
                return terraformer.island_web_world(x, y, tile, map)
            end)
    }
}

-- Tile and Autoplace
local function make_water_autoplace_settings(max_elevation, influence)
    local elevation = noise.var("abyss-elevation")
    local fitness = max_elevation - elevation
    -- Adjust fitness to allow higher-influence (usually deeper) water to override shallower water,
    -- even at elevations where they both have >0 fitness
    local adjusted_fitness = influence * noise.min(fitness, 1)
    return {
        -- If fitness is < 0, probability will be -infinity,
        -- so that water doesn't override the default walkable tile (in case no other tile is placed).
        -- Otherwise probability is adjusted_fitness:
        probability_expression = noise.min(fitness * math.huge, adjusted_fitness)
    }
end

local abyss_tile = Util.copy(data.raw.tile["out-of-map"])
abyss_tile.name = "abyss"
abyss_tile.order = "aa"
abyss_tile.autoplace = make_water_autoplace_settings(0, 300)
abyss_tile.transition_merges_with_tile = "out-of-map"
abyss_tile.map_color = { 0.1, 0.1, 0.1 }
-- collision mask is set in final-fixes, for compatibility with hovercraft mod
-- Util.remove_from_table(abyss_tile.collision_mask, "water-tile")
data:extend { abyss_tile }


-- Preset
------------------------------------------------------------------------
local preset = { order = "a" }

local autoplace_controls = {}
for resource_name, resource in pairs(data.raw.resource) do
    if resource.autoplace then
        autoplace_controls[resource_name] = { frequency = 2, size = 0.5 }
    end
end
autoplace_controls.trees = { frequency = 0.8, size = 0.85 }

preset.basic_settings = {
    autoplace_controls = autoplace_controls,
    terrain_segmentation = 3,
    cliff_settings = { richness = 50, cliff_elevation_interval = 0.1 }
}

local preset_collection = data.raw["map-gen-presets"].default
preset_collection["shattered-land"] = preset
preset_collection.default.order = "ab"


-- Menu Simulations
------------------------------------------------------------------------
local menu_simulations = data.raw["utility-constants"]["default"].main_menu_simulations

local function edit_simulation(name, script_name)
    if menu_simulations[name] then
        menu_simulations[name].save = "__shattered-lands__/menu-simulations/menu-simulation-" ..
        script_name .. "-edited.zip"
    end
end
for name, script_name in pairs({
    artillery = "artillery",
    big_defense = "big-defense",
    burner_city = "burner-city",
    chase_player = "chase-player",
    logistic_robots = "logistic-robots",
    solar_power_construction = "solar-power-construction",
    spider_ponds = "spider-ponds",
    nuclear_power = "nuclear-power",
    early_smelting = "early-smelting",
    train_junction = "train-junction",
    mining_defense = "mining-defense",
    oil_pumpjacks = "oil-pumpjacks",
    brutal_defeat = "brutal-defeat"
}) do
    edit_simulation(name, script_name)
end


-- Integration of other mods
------------------------------------------------------------------------

-- Hovercraft
local hovercraft_tech = data.raw.technology["hcraft-tech"]
if hovercraft_tech then
    hovercraft_tech.prerequisites = {
        "automobilism",
        "fluid-handling"
    }
    hovercraft_tech.unit =
    {
        count = 50,
        ingredients =
        {
            { name = "automation-science-pack", amount = 1 },
            { name = "logistic-science-pack", amount = 1 }
        },
        time = 30
    }
    data.raw.recipe["hcraft-recipe"].ingredients = {
        { name = "steel-plate",      amount = 5 },
        { name = "iron-gear-wheel",  amount = 8 },
        { name = "engine-unit",      amount = 10 },
        { name = "electronic-circuit", amount = 10 }
    }
end


-- Teleporter
local bulk_tele_tech = data.raw.technology["bulkteleport-tech1"]
if bulk_tele_tech then
    bulk_tele_tech.prerequisites = {
        "electric-energy-accumulators",
        "circuit-network",
        "utility-science-pack",
    }
    bulk_tele_tech.unit =
    {
        count = 500,
        ingredients =
        {
            { name = "automation-science-pack", amount = 1 },
            { name = "logistic-science-pack", amount = 1 },
            { name = "chemical-science-pack", amount = 1 },
            { name = "utility-science-pack",  amount = 1 },
            { name = "production-science-pack", amount = 1 },
        },
        time = 30
    }
    local bulk_tele_tech2 = data.raw.technology["bulkteleport-tech2"]
    bulk_tele_tech2.enabled = false
    bulk_tele_tech2.hidden = true

    local bulk_teleporter_sender_buffer = data.raw.container["bulkteleport-buffer-send1"]
    bulk_teleporter_sender_buffer.inventory_size = 40
    local bulk_teleporter_receiver_buffer = data.raw.container["bulkteleport-buffer-recv1"]
    bulk_teleporter_receiver_buffer.inventory_size = 60
    local bulk_teleporter_energizer_send = data.raw.furnace["bulkteleport-energizer-send1"]
    bulk_teleporter_energizer_send.energy_usage = "60MW" -- should use 10MW per belt
    local bulk_teleporter_energizer_receive = data.raw.furnace["bulkteleport-energizer-recv1"]
    bulk_teleporter_energizer_receive.energy_usage = "60MW"
end


-- Lex's Airships
-- unlock the airplanes later and make the recipes more like spidertron recipes (higher tech items, lower amounts)
local jet_engine_tech = data.raw.technology["lex-jet-engine"]
if jet_engine_tech then
    local gunship_tech = data.raw.technology["lex-flying-gunship-ships"]
    local cargo_ship_tech = data.raw.technology["lex-flying-cargo-ships"]
    local heavy_gunship_tech = data.raw.technology["lex-flying-heavyship-ships"]
    local spidertron_tech = data.raw.technology.spidertron
    if jet_engine_tech then
        jet_engine_tech.unit = {
            count = 200,
            ingredients =
            {
                { name = "automation-science-pack", amount = 1 },
                { name = "logistic-science-pack", amount = 1 },
                { name = "chemical-science-pack", amount = 1 },
                { name = "utility-science-pack",  amount = 1 },
            },
            time = 30
        }
    end
    if cargo_ship_tech then
        cargo_ship_tech.unit = {
            count = 300,
            ingredients =
            {
                { name = "automation-science-pack", amount = 1 },
                { name = "logistic-science-pack", amount = 1 },
                { name = "chemical-science-pack", amount = 1 },
                { name = "utility-science-pack",  amount = 1 },
            },
            time = 30
        }
    end
    if gunship_tech then
        gunship_tech.unit = {
            count = 600,
            ingredients =
            {
                { name = "automation-science-pack", amount = 1 },
                { name = "logistic-science-pack", amount = 1 },
                { name = "chemical-science-pack", amount = 1 },
                { name = "military-science-pack", amount = 1 },
                { name = "utility-science-pack",  amount = 1 },
            },
            time = 30
        }
    end
    if heavy_gunship_tech then heavy_gunship_tech.unit.time = 60 end
    if spidertron_tech then
        spidertron_tech.unit =
        {
            count = 500,
            ingredients =
            {
                { name = "automation-science-pack", amount = 1 },
                { name = "logistic-science-pack", amount = 1 },
                { name = "chemical-science-pack", amount = 1 },
                { name = "utility-science-pack",  amount = 1 },
                { name = "production-science-pack", amount = 1 }
            },
            time = 30
        }
    end
    local flying_cargo_recipe = data.raw.recipe["lex-flying-cargo"]
    flying_cargo_recipe.ingredients = {
        { name="raw-fish",              amount=1 },
        { name="rocket-control-unit",   amount=10 },
        { name="low-density-structure", amount=100 },
        { name="flying-robot-frame",    amount=30 },
        { name="radar",                 amount=2 }
    }
    flying_cargo_recipe.normal.ingredients = flying_cargo_recipe.ingredients
    table.insert(cargo_ship_tech.prerequisites, "rocket-control-unit")
    local flying_gunship_recipe = data.raw.recipe["lex-flying-gunship"]
    flying_gunship_recipe.ingredients = {
        { name="raw-fish",              amount=1 },
        { name="low-density-structure", amount=20 },
        { name="steel-plate",           amount=100 },
        { name="flying-robot-frame",    amount=10 },
        { name="submachine-gun",        amount=2 },
        { name="rocket-launcher",       amount=2 },
        { name="radar",                 amount=1 }
    }
    flying_gunship_recipe.normal.ingredients = flying_gunship_recipe.ingredients
    local heavy_gunship_recipe = data.raw.recipe["lex-flying-heavyship"]
    heavy_gunship_recipe.ingredients = {
        { name="raw-fish",              amount=1 },
        { name="low-density-structure", amount=50 },
        { name="steel-plate",           amount=250 },
        { name="flying-robot-frame",    amount=20 },
        { name="rocket-launcher",       amount=10 },
        { name="tank",                  amount=2 },
        { name="radar",                 amount=2 }
    }
    heavy_gunship_recipe.normal.ingredients = heavy_gunship_recipe.ingredients
end

-- Spidertron Patrols
-- data.raw["spider-vehicle"]["sp-spiderling"] = nil
-- create_spidertron{
--     name = "sp-spiderling",
--     scale = 0.7,
--     leg_scale = 0.75, -- relative to scale
--     leg_thickness = 1.2, -- relative to leg_scale
--     leg_movement_speed = 0.62
-- }
-- create_spidertron{
--     name = "sp-spiderling",
--     scale = 0.8,
--     leg_scale = 0.9, -- relative to scale
--     leg_thickness = 1.2, -- relative to leg_scale
--     leg_movement_speed = 0.8
-- }
local spiderling_recipe = data.raw.recipe["sp-spiderling"]
if spiderling_recipe then
    spiderling_recipe.ingredients = {
        { "raw-fish",              1 },
        { "steel-plate",           50 },
        { "copper-plate",          200 },
        { "effectivity-module-2",  2 },
        { "exoskeleton-equipment", 2 },
        { "radar",                 1 },
    }
end

-- Thrower inserter
-- local thrower_tech = data.raw.technology["thrower-inserter"]
-- if thrower_tech then
--     thrower_tech.prerequisites = {

--     }
-- end



-- Scrap Processing 2
if data.raw.technology["scrap-processing"] then
    data.raw.item["military-scrap"].stack_size = 2000
end
